larq/optimizers_v2.py

```diff
@@ -1,10 +1,151 @@
+import warnings
+
 import tensorflow as tf
 import larq as lq
 
 from larq import utils
 from copy import deepcopy
 
 
+@utils.register_keras_custom_object
+class CaseOptimizer(tf.keras.optimizers.Optimizer):
+    """A case operation for optimizers that each train a subset of a model's variables.
+
+    An optimizer is used to train a variable iff its accompanying predicate evaluates to
+    `True`.
+
+    For each variable, at most one optimizer's predicate may evaluate to `True`. If no
+    optimizer's predicate evaluates to `True` for a variable, it is trained with the
+    `default` optimizer.
+
+    # Arguments
+    pred_opt_pairs: a list of `(tf.keras.optimizers.Optimzer, pred)` pairs, where `pred` 
+        takes one `tf.Variable` as argument and returns `True` if the optimizer should
+        train that variable, e.g. `pred(var) == True`.
+    default: a `tf.keras.optimizers.Optimizer` to be applied to any variable not claimed
+        by any other optimizer.
+    """
+
+    def __init__(self, pred_opt_pairs, default=None, name="optimizer_case"):
```

Comment:
Should we change the signature to
```python
    def __init__(self, *pred_opt_pairs, default=None, name="optimizer_case"):
```
Then we can get rid of an extra list when using the optimizer.

Reply:
I don't think this works with tuples nested inside lists; the different variations of this I tried all ended up failing for either `[(pred, opt)]` or `(pred, opt)`. If you have a way that does work please let me know.

Reply:
I don't see why this would fail. The following works well for me:
```python
def foo(*pred_opt_pairs):
    print(pred_opt_pairs) 

foo((True, "opt1"), (True, "opt2"))
# --> ((True, 'opt1'), (True, 'opt2'))

foo((True, "opt1"))
# --> ((True, 'opt1'),)
```

Reply:
Ah, I was trying tuples nested inside lists instead of just several tuples after each other, but latter works. The only drawback here is that you always have to pass `default=` for the default, or it'll try to unpack that as a tuple too; but I think that's fine. Changed in 2ec9175.